### DISCUSSION TITLE
How to force static dependencies without specifying whole triplet?

### DISCUSSION THREAD

**Discussion by Chlumsky** [microsoft/vcpkg#22099] [Q&A]
I'd like to specify that all of the library dependencies of my program should be built as static libraries either in the JSON manifest or in the CMakeLists. However, I don't want to set the whole triplet such as `x86-windows-static` because I still want it to automatically work on any OS / platform I'm using at that moment. How would I achieve this?

**Comment by Osyotr** [comment]
It was covered in this example:
https://github.com/microsoft/vcpkg/blob/master/docs/examples/overlay-triplets-linux-dynamic.md#overriding-default-triplets

**Reply by Chlumsky** [reply]
Either that doesn't solve the problem I described or I don't understand how. Can you explain specifically how I can use it to achieve my goal?

**Reply by Osyotr** [reply]
As of today, it's impossible to specify linkage for libraries build with vcpkg in JSON manifest or in CMakeLists.txt.
If you don't want to explicitly specify triplet, you have to explicitly specify `overlay-triplets/` folder (copy of files in `vcpkg/triplets/` but with `-static` omitted).
Why not specify `VCPKG_TARGET_TRIPLET` it in `CMakePresets.json`?

**Reply by Chlumsky** [reply]
I am not looking to set this globally for my computer, but for the dependencies of a specific repository (cloned on multiple computers). Why is this not possible? It seems like it shouldn't be a problem and would be quite useful - for example what if I wanted to link some libraries statically and some dynamically?

**Reply by Neumann-A** [reply]
you want the above (to not add custom triplets to the vcpkg triplet dir) + https://github.com/microsoft/vcpkg/blob/master/docs/users/triplets.md#per-port-customization

**Reply by Chlumsky** [reply]
Adding `set(VCPKG_LIBRARY_LINKAGE static)` to CMakeLists was one of the first things I tried but it had zero effect. What do you mean by "the above"?

**Reply by Neumann-A** [reply]
> to CMakeLists was one of the first things I tried but it had zero effect.

You didn't read the docs correctly. You don't add that to your CMakeLists.txt..... you set it in your triplet file.

**Reply by Chlumsky** [reply]
Yes I thought so, but you didn't read the title of the question correctly. I can't set anything in a triplet file if I don't know the rest of the triplet (platform and architecture) and it's supposed to work for all of them.

**Reply by autoantwort** [reply]
That is currently not possible, you have to create an own triplet for that. If you need this for every architecture, you have to create a custom triplet for every architecture. There are currently no other solutions.

**Reply by Chlumsky** [reply]
Is this something that might be added in the future or is there a conceptual problem that makes it not possible / undesirable?

**Reply by Neumann-A** [reply]
> undesirable

- CMakeLists.txt describes how to build your project not your dependencies
- vcpkg manifest (vcpkg.json) list with your vcpkg dependencies which your CMakeLists.txt plans to consume somehow (vcpkg generally does not know how you consume). 
- vcpkg triplet -> file with vcpkg internal settings; i.e. how is the library linkage of xy or what the target architecture is etc. 
- vcpkg portfile.cmake -> how will xy be build by vcpkg. 


If you want the behavior from your CMakeLists.txt. You can try to generate a triplet before `project()` is called and use that for vcpkg.

**Reply by autoantwort** [reply]
It is a known problem. I think they want to improve it, but they currently don't know how (because they haven't started thinking about it yet).

**Reply by Chlumsky** [reply]
> CMakeLists.txt describes how to build your project not your dependencies

"Statically" is a description of how my project should be built. I am also OK with specifying this in the vcpkg.json, which would make the most sense to me because specifying if a dependency should be static / dynamic is just like selecting the version.

I know that I can set the triplet at the beginning of the CMake but not the static part by itself which is what I've been getting at from the beginning. ~Also, editting the triplet files doesn't work for me because I don't have an instance of Vcpkg checked out specifially for this project.~

**Reply by Neumann-A** [reply]
> because specifying if a dependency should be static / dynamic is just like selecting the version.

nope it is not. A version is always a minimum (or exact) build requirement (due to e.g. API) while static / dynamic is most of the time just a preference or license based decision (unless there is some sort of global shared state which forces dynamic linkage).  

> I know that I can set the triplet at the beginning of the CMake but not the static part by itself 
> I don't have an instance of Vcpkg checked out specifially

overlay triplets (`--overlay-triplets=<path_to_folder_with_triplet>`)! Those were already mentioned here. 
And as I said you can generate the triplet file and just point to it.

**Reply by DDoS** [reply]
I have the same issue as OP. I just want to extend the default triplet, not create a new one. I don't want to have to redefine a triplet for every platform that might be used, I just want to make some customizations on some platforms. My goal would be to have a triplet that looks like:
```cmake
# Initialize my custom triplet using the default detected by vcpkg
include(${VCPKG_DEFAULT_TRIPLET_PATH})

# Customize the default triplet
if(UNIX)
    set(VCPKG_CMAKE_CONFIGURE_OPTIONS
	    "-DCMAKE_C_VISIBILITY_PRESET=hidden"
	    "-DCMAKE_CXX_VISIBILITY_PRESET=hidden")
endif()
# Whatever else I want.
``` 
This if for a CMake project where I want to tweak how the dependencies are built.
